timemachine/cpp/src/local_md_config.cu

```diff
@@ -0,0 +1,197 @@
+    // Simply reseeding does NOT produce identical results
+    curandErrchk(curandSetGeneratorOffset(cr_rng_, 0));
+
+    // Set the array to all N, which indicates to ignore that idx
+    k_initialize_array<unsigned int>
+        <<<ceil_divide(N_, warp_size), warp_size, 0, stream>>>(N_, d_free_indices_.data, N_);
+    gpuErrchk(cudaPeekAtLastError());
+
+    // Generate values between (0, 1.0]
+    curandErrchk(curandGenerateUniform(cr_rng_, probability_buffer_.data, round_up_even(N_)));
+
+    std::mt19937 rng;
+    rng.seed(seed);
+    std::uniform_int_distribution<unsigned int> random_dist(0, local_idxs.size() - 1);
+
+    unsigned int reference_idx = local_idxs[random_dist(rng)];
+
+    const double kBT = BOLTZ * temperature;
+    // Select all of the particles that will be free
+    k_log_probability_selection<float><<<ceil_divide(N_, warp_size), warp_size, 0, stream>>>(
+        N_, kBT, radius, k, reference_idx, d_x_t, d_box_t, probability_buffer_.data, d_free_indices_.data);
+    gpuErrchk(cudaPeekAtLastError());
+
+    this->_setup_free(reference_idx, radius, k, stream);
+}
+
+void LocalMDConfig::_setup_free(
+    const unsigned int reference_idx, const double radius, const double k, const cudaStream_t stream) {
+    const int tpb = warp_size;
+
+    LessThan select_op(N_);
+
+    // Partition the free indices into the row indices
+    gpuErrchk(cub::DevicePartition::If(
+        d_temp_storage_buffer_->data,
+        temp_storage_bytes_,
+        d_free_indices_.data,
+        d_row_idxs_.data,
+        num_selected_buffer_.data,
+        N_,
+        select_op,
+        stream));
+
+    gpuErrchk(cudaMemcpyAsync(
+        p_num_selected_.data,
+        num_selected_buffer_.data,
+        1 * sizeof(*p_num_selected_.data),
+        cudaMemcpyDeviceToHost,
+        stream));
+    gpuErrchk(cudaStreamSynchronize(stream));
+
+    const int num_row_indices = p_num_selected_.data[0];
+    const int num_col_indices = N_ - num_row_indices;
+
+    if (num_row_indices == 0) {
+        throw std::runtime_error("LocalMDConfig setup has no free particles selected");
+    }
+
+    // The reference particle will always be in the column indices
+    if (num_row_indices == N_ - 1) {
+        fprintf(stderr, "LocalMDConfig setup has entire system selected\n");
+    }
+
+    k_construct_bonded_params<<<ceil_divide(num_row_indices, tpb), tpb, 0, stream>>>(
+        num_row_indices, N_, reference_idx, k, 0.0, radius, d_row_idxs_.data, restraints_.data, bond_params_.data);
+    gpuErrchk(cudaPeekAtLastError());
+
+    // Setup the flat bottom restraints
+    bound_restraint_->set_params_device(std::vector<int>({num_row_indices, 3}), bond_params_.data, stream);
+    restraint_->set_bonds_device(num_row_indices, restraints_.data, stream);
+
+    // Set the nonbonded potential to compute forces of free particles
+    set_nonbonded_potential_idxs(nonbonded_bp_->potential, num_row_indices, d_row_idxs_.data, stream);
```

Comment:
tiny nit: mixed convention with "indices" and "idxs" (might have been present before)